core-primitives/attestation-handler/src/collateral.rs

```diff
@@ -0,0 +1,155 @@
+	/// its members have a value that is not nullptr
+	pub unsafe fn from_c_type(c: &sgx_ql_qve_collateral_t) -> Self {
+		let pck_crl_issuer_chain = std::slice::from_raw_parts(
+			c.pck_crl_issuer_chain as *const u8,
+			c.pck_crl_issuer_chain_size as usize,
+		)
+		.to_vec();
+		let root_ca_crl =
+			std::slice::from_raw_parts(c.root_ca_crl as *const u8, c.root_ca_crl_size as usize)
+				.to_vec();
+		let pck_crl =
+			std::slice::from_raw_parts(c.pck_crl as *const u8, c.pck_crl_size as usize).to_vec();
+		let tcb_info_issuer_chain = std::slice::from_raw_parts(
+			c.tcb_info_issuer_chain as *const u8,
+			c.tcb_info_issuer_chain_size as usize,
+		)
+		.to_vec();
+		let tcb_info =
+			std::slice::from_raw_parts(c.tcb_info as *const u8, c.tcb_info_size as usize).to_vec();
+		let qe_identity_issuer_chain = std::slice::from_raw_parts(
+			c.qe_identity_issuer_chain as *const u8,
+			c.qe_identity_issuer_chain_size as usize,
+		)
+		.to_vec();
+		let qe_identity =
+			std::slice::from_raw_parts(c.qe_identity as *const u8, c.qe_identity_size as usize)
+				.to_vec();
+		SgxQlQveCollateral {
+			version: c.version,
+			tee_type: c.tee_type,
+			pck_crl_issuer_chain,
+			root_ca_crl,
+			pck_crl,
+			tcb_info_issuer_chain,
+			tcb_info,
+			qe_identity_issuer_chain,
+			qe_identity,
+		}
+	}
+
+	pub fn dump_to_disk(&self) {
+		Self::write_data_to_disk("pck_crl_issuer_chain", &self.pck_crl_issuer_chain);
+		Self::write_data_to_disk("root_ca_crl", &self.root_ca_crl);
+		Self::write_data_to_disk("pck_crl", &self.pck_crl);
+		Self::write_data_to_disk("tcb_info_issuer_chain", &self.tcb_info_issuer_chain);
+		Self::write_data_to_disk("tcb_info", &self.tcb_info);
+		Self::write_data_to_disk("qe_identity_issuer_chain", &self.qe_identity_issuer_chain);
+		Self::write_data_to_disk("qe_identity", &self.qe_identity);
+	}
+
+	/// Returns the tcb_info split into two parts: json_data and signature
+	pub fn get_tcb_info_split(&self) -> Option<(String, Vec<u8>)> {
+		let (json_data, signature) =
+			Self::separate_json_data_and_signature("tcbInfo", &self.tcb_info)?;
+		match hex::decode(signature) {
+			Ok(hex_signature) => Some((json_data, hex_signature)),
+			Err(_) => None,
+		}
+	}
+
+	/// Returns the tcb_info split into two parts: json_data and signature
+	pub fn get_quoting_enclave_split(&self) -> Option<(String, Vec<u8>)> {
+		let (json_data, signature) =
+			Self::separate_json_data_and_signature("enclaveIdentity", &self.qe_identity)?;
+		match hex::decode(signature) {
+			Ok(hex_signature) => Some((json_data, hex_signature)),
+			Err(_) => None,
+		}
+	}
+
+	/// Separates the actual data part from the signature for an Intel collateral in JSON format
+	/// Returns the data part and signature as a pair
+	fn separate_json_data_and_signature(data_name: &str, data: &[u8]) -> Option<(String, String)> {
+		let json = String::from_utf8_lossy(data);
+		let value: serde_json::Value = serde_json::from_str(&json).ok()?;
```

Comment:
On the call site data is often a C terminated string, and I encountered an issue, where the terminating `\0` was still there and serde failed with unexpected ending character (a simple `println!()` did not show it, it needed to be debug formatted).

For this reason I suggest this:
```suggestion
        let json = CStr::from_bytes_with_nul(data)
			.expect("CStr::from_bytes_with_nul failed")
			.to_str()?;
		let value: serde_json::Value = serde_json::from_str(&json).ok()?;
```

Reply:
Can you explain, why it is safe to put an expect here instead of returning an error?

Reply:
~Of course~:
- `from_bytes_with_nul().expect()` is safe, because `SgxQlQveCollateral` is the Rust version of `sgx_ql_qve_collateral_t`, [which states the following:](https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf)
    ```
   /** This is the data provided to the quote verifier by the verifying platform
   software. They are NULL terminated strings. [...] */
    ```
-  `to_str().expect()` ~is safe, because~ I could not find an explicit guarantee regarding the fields always being valid UTF-8 sequences. -> The second `.expect()` should be `.ok()?` instead, right?

Nice catch @clangenb, sorry for assuming too much!

Reply:
Thanks for the explanation. However, I am still unsure. We must never panic in the enclave, as this might leak sensitive data. So I prefer that the function doesn't assume that it is called in the correct context. I prefer to be defensive here and also handle the error case. In general, I only allow `expect`, when the context is static, and it does really not allow panics.

Reply:
Alright, I pushed the changes.